gammapy/maps/hpxnd.py

```diff
@@ -452,6 +453,47 @@ def smooth(self, width, kernel="gauss"):
 
         return self._init_copy(data=smoothed_data)
 
+    def convolve_wcs(self, kernel, use_fft=True, **kwargs):
+        """
+        Convolve map with a WCS kernel.
+
+        It projects the map into a WCS geometry, convolves with a WCS kernel and
+        projects back into the initial Healpix geometry.
+
+        If the kernel is two dimensional, it is applied to all image planes likewise.
+        If the kernel is higher dimensional it must match the map in the number of
+        dimensions and the corresponding kernel is selected for every image plane.
+
+        Parameters
+        ----------
+        kernel : `~gammapy.irf.PSFKernel` or `numpy.ndarray`
+            Convolution kernel. Must have the same geometry as
+            `self.to_wcs()`.
+        use_fft : bool
+            Use `scipy.signal.fftconvolve` if True (default)
+            and `ndi.convolve` otherwise.
+        kwargs : dict
+            Keyword arguments passed to `scipy.signal.fftconvolve` or
+            `ndi.convolve`.
+
+        Returns
+        -------
+        map : `HpxNDMap`
+            Convolved map.
+        """
+        if self.geom.width < 10*u.deg:
+            # Project to WCS and convolve
+            wcs_map  = self.to_wcs(fill_nan = False)
```

Comment:
I think `.to_wcs()` takes a pre-computed mapping as well. There might be a performance improvement if the `HpxToWcsMapping` is created outside and re-used for both cases i.e. `.to_wcs()` as well as the later inverse operation.

Reply:
I think this even offers the possibly to adjust the WCS geometry used for the conversion and convolution, to the WCS geometry of the kernel, that is passed in. This eliminates the need to compute a matching kernel outside. I haven't thought this through completely, but I think it works...

Reply:
oh, you are right! that is a very good point! I'll try that out

Reply:
It works but I am seeing a lot more issues with artifacts that only go away if the wcs_geom binning is very fine. The nice thing about using `.to_wcs()` is that it guaranteed the oversampling... I guess it could be fixed by requiring a minimum wcs pixel size ratio to hpx pixel size...
```
%%time
energy = MapAxis.from_bounds(1,100, unit='TeV', nbin=2, name='energy')

nside = 1024
hpx_geom = HpxGeom.create(nside=nside,
                         axes=[energy],
                         region='DISK(0,0,2.5)',
                         )

hpx_map = Map.from_geom(hpx_geom)
hpx_map.set_by_coord((0,0, [2,90]), 1)

wcs_geom = WcsGeom.create(width=5,
                          binsz = 0.05,
                         axes = [energy])
kernel = PSFKernel.from_gauss(wcs_geom, 0.5*u.deg)
convolved_map = hpx_map.convolve_wcs(kernel)
convolved_map.plot_interactive(add_cbar=True)
```
![image](https://user-images.githubusercontent.com/22521727/116532278-59f06700-a8e0-11eb-983c-b6075bbca91a.png)

Reply:
Thanks for check! Indeed we should require the oversampling to avoid the artefacts. As an educated guess, I'd say an oversampling of >2 should be sufficient. So +1 to just include the check and document it.